pkg/source/saddle/pool_simulator.go

```diff
@@ -0,0 +1,268 @@
+package saddle
+
+import (
+	"encoding/json"
+	"errors"
+	"math/big"
+	"strings"
+
+	"github.com/KyberNetwork/kyberswap-dex-lib/pkg/entity"
+	"github.com/KyberNetwork/kyberswap-dex-lib/pkg/source/pool"
+	constant "github.com/KyberNetwork/kyberswap-dex-lib/pkg/util/bignumber"
+	utils "github.com/KyberNetwork/kyberswap-dex-lib/pkg/util/bignumber"
+)
+
+type Pool struct {
+	pool.Pool
+	Multipliers []*big.Int
+	// extra fields
+	InitialA           *big.Int
+	FutureA            *big.Int
+	InitialATime       int64
+	FutureATime        int64
+	AdminFee           *big.Int
+	DefaultWithdrawFee *big.Int
+	LpToken            string
+	LpSupply           *big.Int
+	gas                Gas
+}
+
+func NewPoolSimulator(entityPool entity.Pool) (*Pool, error) {
+	var staticExtra StaticExtra
+	if err := json.Unmarshal([]byte(entityPool.StaticExtra), &staticExtra); err != nil {
+		return nil, err
+	}
+
+	var extra Extra
+	if err := json.Unmarshal([]byte(entityPool.Extra), &extra); err != nil {
+		return nil, err
+	}
+
+	numTokens := len(entityPool.Tokens)
+	tokens := make([]string, numTokens)
+	reserves := make([]*big.Int, numTokens)
+	multipliers := make([]*big.Int, numTokens)
+	for i := 0; i < numTokens; i += 1 {
+		tokens[i] = entityPool.Tokens[i].Address
+		reserves[i] = utils.NewBig10(entityPool.Reserves[i])
+		multipliers[i] = utils.NewBig10(staticExtra.PrecisionMultipliers[i])
+	}
+
+	swapFee := utils.NewBig10(extra.SwapFee)
+
+	// only have withdrawFee in saddle v1, default to 0
+	defaultWithdrawFee := utils.NewBig10(extra.DefaultWithdrawFee)
+	if defaultWithdrawFee == nil {
+		defaultWithdrawFee = constant.ZeroBI
```

Comment:
looks like there is a mistake here
```suggestion
		defaultWithdrawFee = bignumber.ZeroBI
```

Reply:
oh, you alias `bignumber` to `constant`.
It's a bit confusing.

I don't have a strong opinion on this, just make sure to keep it consistent

Reply:
yes, sorry for the confusing, I think using alias will reduce the number of change need to be made